### PR TITLE
feat: [EXT-3648][EXT-4380] Improve sdk.field types

### PR DESCRIPTION
**Pull request by vikaskumr**
# Purpose of PR
Narrow `field` types, particularly properties `type`, `linkType`, `items`.

# Result
image

### CODE REVIEW

File: lib/field-locale.ts [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
 
 export default class FieldLocale implements FieldAPI {
   id: string
   locale: string
-  type: string
+  type: FieldLocaleType
   required: boolean
   validations: any[]
-  items?: Items
+  items: never
```
Comment (colomolo) [inline lib/field-locale.ts:13]: Why it's `never`?

File: lib/field-locale.ts [hunk 2]
```diff
@@ -1,15 +1,16 @@
...
-import { FieldAPI, FieldInfo, Items, SerializedJSONValue } from './types'
+import { FieldAPI, FieldInfo, SerializedJSONValue } from './types'
+import { FieldLocaleType } from './types/utils'
 import { ValidationError } from './types/validation-error'
 
 export default class FieldLocale implements FieldAPI {
   id: string
   locale: string
-  type: string
+  type: FieldLocaleType
```
Comment (colomolo) [inline lib/field-locale.ts:10]: Naming could be better. This is a field type, right? And no need to use the word 'type' in the name of type.

File: lib/field-locale.ts [hunk 3]
```diff
@@ -1,15 +1,16 @@
 import { Channel } from './channel'
 import { MemoizedSignal } from './signal'
-import { FieldAPI, FieldInfo, Items, SerializedJSONValue } from './types'
+import { FieldAPI, FieldInfo, SerializedJSONValue } from './types'
+import { FieldLocaleType } from './types/utils'
```
Comment (colomolo) [inline lib/field-locale.ts:4]: Feels like this should be in './types'

File: lib/types/utils.ts [hunk 4]
```diff
@@ -60,12 +60,18 @@ export interface EntrySys extends ContentEntitySys {
   automationTags: Link<'Tag'>[]
 }
 
+type FieldType = 'Symbol' | 'Number' | 'Array' | 'Reference'
 export interface Items {
-  type: string
-  linkType?: string
+  type: FieldType
```
Comment (colomolo) [inline lib/types/utils.ts:65]: Can Items be of type 'Array'? Worth checking.

File: lib/types/utils.ts [hunk 5]
```diff
@@ -60,12 +60,18 @@ export interface EntrySys extends ContentEntitySys {
...
+  linkType?: FieldType extends 'Reference' ? 'Entry' | 'Asset' : never
   validations?: ContentTypeFieldValidation[]
 }
 
+export type FieldLocaleType =
+  | { type: 'Symbol' }
+  | { type: 'Number' }
+  | { type: 'Array'; items: Items }
```
Comment (colomolo) [inline lib/types/utils.ts:73]: I think 'Reference' should be also here?

**Review by colomolo — COMMENTED** [review]
In general, I would take another approach. We have three distinctive types with common properties and individual properties. So we can use discriminated unions to define them (see 'Discriminated union' [here](https://blog.logrocket.com/understanding-discriminated-union-intersection-types-typescript/)):
```typescript
interface FieldBase {
    id: string
    name: string
    locale: string
    required: boolean
    validations: ContentTypeFieldValidation[]
    value: any
    isDisabled: boolean
    schemaErrors: ValidationError[]
}

interface SymbolField extends FieldBase {
    type: 'Symbol'
}

interface NumberField extends FieldBase {
    type: 'Number'
}

interface ArrayField extends FieldBase {
    type: 'Array'
    items: Items
}

interface ReferenceField extends FieldBase {
    type: 'Reference'
    linkType: 'Entry' | 'Asset'
}

type FieldInfo = SymbolField | NumberField | ArrayField | ReferenceField
```

Not sure if this will cause us problems down the line, but should be something along these lines.

File: lib/types/utils.ts [hunk 6]
```diff
@@ -60,11 +60,23 @@ export interface EntrySys extends ContentEntitySys {
...
+  type: FieldType
   validations?: ContentTypeFieldValidation[]
 }
+export interface ReferenceItems extends ItemsBase {
+  type: 'Reference'
+  linkType: 'Entry' | 'Asset'
+}
+
+export interface PrimitiveItems extends ItemsBase {
+  type: 'Symbol' | 'Number' | 'Array'
```
Comment (colomolo) [inline lib/types/utils.ts:74]: Not sure ifan item can be of 'Array' type. Needs check.

File: package-lock.json
Comment (andipaetzold) [inline package-lock.json:1]: Please revert

**Review by andipaetzold — CHANGES_REQUESTED** [review]
You have now properly typed the internal connect message. Any reason for not typing the external api (FieldAPI & EntryFieldAPI) as well?

**Comment by colomolo** [comment]
Maybe not the best naming: `BasicEntryFieldInfo`, `ExhaustiveEntryFieldAPI`, suggestions are more than welcome.

File: lib/field-locale.ts [hunk 7]
```diff
@@ -23,8 +26,12 @@ export default class FieldLocale implements FieldAPI {
     this.type = info.type
     this.required = info.required
     this.validations = info.validations
-    this.items = info.items
-
+    if (info.type === 'Array') {
```
Comment (andipaetzold) [inline lib/field-locale.ts:29]: Does the condition change anything? If the type is not array, `items` is anyway initialized with `undefined`. So, this only prevents issues if we send an invalid connect message, right?

File: lib/types/index.ts [hunk 8]
```diff
@@ -93,6 +93,7 @@ export type {
   EntrySys,
   ContentEntitySys,
   ContentEntityType,
+  FieldType,
```
Comment (andipaetzold) [inline lib/types/index.ts:96]: Also export `FieldLinkType`?

**Review by andipaetzold — APPROVED** [review]
Must be merged as `feat:` or `fix:` to trigger a new release

About the name: I have no better name myself, but as we are not exporting that type this should not be a blocker. Maybe a native speaker has a better name.

File: lib/field-locale.ts (hunk 7 above)
Comment (colomolo) [reply lib/field-locale.ts:29]: Typescript doesn't like if we do it in a straightforward way:
```js
this.items = info.items
this.linkType = info.linkType
```
> Property 'items' does not exist on type 'BasicFieldInfo'.ts(2339)

We can do it like this:
```js
this.items = (info as ArrayFieldAPI).items
```
But I wouldn't export `ArrayFieldAPI` just for this one case.

So yeah, just an extra check.

Comment (colomolo) [reply lib/field-locale.ts:29]: Btw, we can delete `items` and `linkType` if `type` is not an Array or Link 🤔 Wdyt?

File: lib/types/index.ts (hunk 8 above)
Comment (colomolo) [reply lib/types/index.ts:96]: Do you think we should expose it?

Comment (andipaetzold) [reply lib/types/index.ts:96]: We export types (`EntryFieldAPI`) that use this type, so yes.

Comment (colomolo) [reply lib/types/index.ts:96]: Done

**Review by mgoudy91 — APPROVED** [review]
Still coming up to speed but having seen the previous issues with typings I think this is a clean solution.

Also, I think the names are fine and fit the rest of the implementation well enough and wouldn't worry about renaming.

File: lib/types/utils.ts [hunk 9]
```diff
@@ -60,12 +60,35 @@ export interface EntrySys extends ContentEntitySys {
...
+  | 'Symbol'
+  | 'Text'
+  | 'RichText'
+  | 'Number'
+  | 'Integer'
+  | 'Array'
+  | 'Link'
+  | 'Object'
+  | 'Date'
+  | 'Location'
```
Comment (gabiayako) [inline lib/types/utils.ts:73]: Isn't the type Boolean necessary here?

Comment (colomolo) [reply lib/types/utils.ts:73]: Good catch! I will add it.

Comment (colomolo) [reply lib/types/utils.ts:73]: https://github.com/contentful/ui-extensions-sdk/pull/1541

**Pull request merged**